distributed/scheduler.py

```diff
@@ -1975,7 +1966,50 @@ def new_task(self, key: str, spec: object, state: str) -> TaskState:
     # State Transitions #
     #####################
 
-    def _transition(self, key, finish: str, *args, **kwargs):
+    def _transition_dispatch(self, start: str, finish: str, *args, **kwargs):
+        start_finish: str = sys_intern(f"{start}_{finish}".replace("-", "_"))
```

Comment:
I'd be interested in a speed test of this approach versus `hash`/dict lookup or actual `strcmp` (which is what I assume cython does when both operands are `char*`). For the latter, you don't need the string manipulation to make the compound, and the if statements can be grouped.